src/auth/mod.rs

```diff
@@ -0,0 +1,128 @@
+pub(super) use crate::actix_web::Result as ActixResult;
+use crate::actix_web::{error::ErrorUnauthorized, HttpRequest};
+use actix_web::http::header::HeaderMap;
+
+pub mod jwt;
+
+#[derive(Clone)]
+pub struct AuthHeader {
+    field: &'static str,
+    token_bound: (Option<&'static str>, Option<&'static str>),
+}
+
+impl AuthHeader {
+    /// return None if value is invalid or can't be parsed
+    pub fn new(field: &'static str, value: &'static str) -> Option<Self> {
+        let mut not_token = value.trim().split("{token}");
+        let token_bound = (
+            not_token.next().filter(|s| !s.is_empty()),
+            match not_token.next() {
+                None => return None,
+                Some(s) if s.is_empty() => None,
+                Some(s) => Some(s),
+            },
+        );
+        Some(Self { field, token_bound })
+    }
+}
+
+impl Default for AuthHeader {
+    fn default() -> Self {
+        AuthHeader::new("Authorization", "Bearer {token}").expect("has {token}")
+    }
+}
+
+#[derive(Clone)]
+pub enum Auth {
+    JWT {
+        /** Header where the authentication token reside.\n
+        The format value is always be `... {token} ...`.\n
+        Default is `Authorization: Bearer {token}` */
+        auth_header: AuthHeader,
+        /** Bytes used for secret.
+        Use std::include_bytes!(from_file) for convinience */
+        signing_secret: &'static [u8],
+        /// default is RS256 (RSA using SHA-256)
+        algorithm: jwt::SignatureAlgorithm,
+        validate: jwt::ClaimCode,
+    },
+    None,
+}
+
+impl Default for Auth {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+impl Auth {
+    pub fn default_jwt_from(signing_secret: &'static [u8]) -> Self {
+        Self::JWT {
+            auth_header: AuthHeader::new("Authorization", "Bearer {token}").expect("has {token}"),
+            validate: jwt::ClaimCode::disable_all(),
+            algorithm: jwt::SignatureAlgorithm::RS256,
+            signing_secret,
+        }
+    }
+
+    pub(crate) fn validate(&self, request: &HttpRequest) -> ActixResult<()> {
+        match self {
+            Self::None => Ok(()),
+            Self::JWT {
+                auth_header: template,
+                validate: claim_code,
+                signing_secret: secret,
+                algorithm,
+            } => {
+                let token = extract_token(template, request.headers())?;
+                claim_code.validate(secret, token, *algorithm)
+            }
+        }
+    }
+}
+
+fn extract_token<'a>(template: &AuthHeader, header: &'a HeaderMap) -> ActixResult<&'a str> {
+    let header_value = header.get(template.field).ok_or_else(|| {
+        let message = ["Missing field '", template.field, "'"].concat();
+        ErrorUnauthorized(message)
+    })?;
+
+    let mut token = header_value.to_str().map_err(|e| ErrorUnauthorized(e.to_string()))?;
+    if let Some(non_token) = template.token_bound.0 {
```

Comment:
why is there `template.token_bound.0` and `template.token_bound.1`? the properties `0` and `1` are weird namings don't you think. Am I missing something here

Reply:
That's how to access tuple in Rust.
Actually, it's more readable if I use [regex crate][] but it will increase the compile time 😂

Hmm, I think it's better if I destructuring assignment 🤔
```rs
let (prefix, suffix) = template.token_bound;
```

[regex crate]: https://docs.rs/regex

Reply:
what the fuckkkkkkkkk really? `.0` and `.1`. Ok I'm forgiving this now

Reply:
Tuple destructuring is preferable, we should discuss more about the convention in Discourse.